plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TypeConverter.java

```diff
@@ -179,25 +191,56 @@ private static org.apache.iceberg.types.Type fromDecimal(DecimalType type)
+            int id = childColumnIdentity
+                    .map(ColumnIdentity::getId)
+                    .orElseGet(() -> nextFieldId.orElseThrow(() -> new IllegalArgumentException("Either a column identity or nextFieldId is expected")).getAndIncrement());
+
+            RowType.Field field = type.getFields().get(fieldIndex);
             String name = field.getName().orElseThrow(() ->
                     new TrinoException(NOT_SUPPORTED, "Row type field does not have a name: " + type.getDisplayName()));
-            fields.add(Types.NestedField.optional(fields.size() + 1, name, toIcebergType(field.getType())));
+            fields.add(Types.NestedField.optional(id, name, toIcebergTypeInternal(field.getType(), childColumnIdentity, nextFieldId)));
         }
         return Types.StructType.of(fields);
     }
 
-    private static org.apache.iceberg.types.Type fromArray(ArrayType type)
+    private static org.apache.iceberg.types.Type fromArray(ArrayType type, Optional<ColumnIdentity> columnIdentity, Optional<AtomicInteger> nextFieldId)
+    {
+        checkExactlyOne(columnIdentity, nextFieldId);
+        Optional<ColumnIdentity> childColumnIdentity = columnIdentity.map(identity -> getOnlyElement(identity.getChildren()));
+        int id = childColumnIdentity
+                .map(ColumnIdentity::getId)
+                .orElseGet(() -> nextFieldId.orElseThrow(() -> new IllegalArgumentException("Either a column identity or nextFieldId is expected")).getAndIncrement());
+        return Types.ListType.ofOptional(id, toIcebergTypeInternal(type.getElementType(), childColumnIdentity, nextFieldId));
+    }
+
+    private static org.apache.iceberg.types.Type fromMap(MapType type, Optional<ColumnIdentity> columnIdentity, Optional<AtomicInteger> nextFieldId)
     {
-        return Types.ListType.ofOptional(1, toIcebergType(type.getElementType()));
+        checkExactlyOne(columnIdentity, nextFieldId);
+        Optional<ColumnIdentity> keyColumnIdentity = columnIdentity.map(column -> column.getChildren().get(0));
+        Optional<ColumnIdentity> valueColumnIdentity = columnIdentity.map(column -> column.getChildren().get(1));
+        int keyId = keyColumnIdentity
+                .map(ColumnIdentity::getId)
+                .orElseGet(() -> nextFieldId.orElseThrow(() -> new IllegalArgumentException("Either a column identity or nextFieldId is expected")).getAndIncrement());
+        int valueId = valueColumnIdentity
+                .map(ColumnIdentity::getId)
+                .orElseGet(() -> nextFieldId.orElseThrow(() -> new IllegalArgumentException("Either a column identity or nextFieldId is expected")).getAndIncrement());
+        return Types.MapType.ofOptional(keyId, valueId, toIcebergTypeInternal(type.getKeyType(), keyColumnIdentity, nextFieldId), toIcebergTypeInternal(type.getValueType(), valueColumnIdentity, nextFieldId));
```

Comment:
too long. Please break it.
```suggestion
        return Types.MapType.ofOptional(
                keyId,
                valueId,
                toIcebergTypeInternal(type.getKeyType(), keyColumnIdentity, nextFieldId),
                toIcebergTypeInternal(type.getValueType(), valueColumnIdentity, nextFieldId));
```